public/app/features/dashboard/components/PanelEditor/PanelEditorTabs.tsx

```diff
@@ -17,77 +17,65 @@ interface PanelEditorTabsProps {
   onChangeTab: (tab: PanelEditorTab) => void;
 }
 
-export class PanelEditorTabs extends PureComponent<PanelEditorTabsProps> {
-  private eventSubs = new Subscription();
+export const PanelEditorTabs: FC<PanelEditorTabsProps> = React.memo(({ panel, dashboard, tabs, onChangeTab }) => {
+  const forceUpdate = useForceUpdate();
+  const styles = useStyles(getStyles);
 
-  componentDidMount() {
-    const { events } = this.props.panel;
-    this.eventSubs.add(events.subscribe(PanelQueriesChangedEvent, this.triggerForceUpdate));
-    this.eventSubs.add(events.subscribe(PanelTransformationsChangedEvent, this.triggerForceUpdate));
-  }
-
-  componentWillUnmount() {
-    this.eventSubs.unsubscribe();
-  }
-
-  triggerForceUpdate = () => {
-    this.forceUpdate();
-  };
-
-  getCounter = (tab: PanelEditorTab) => {
-    const { panel } = this.props;
+  useEffect(() => {
+    const eventSubs = new Subscription();
+    eventSubs.add(panel.events.subscribe(PanelQueriesChangedEvent, forceUpdate));
+    eventSubs.add(panel.events.subscribe(PanelTransformationsChangedEvent, forceUpdate));
+    return () => eventSubs.unsubscribe();
+  }, [panel, forceUpdate]);
 
-    switch (tab.id) {
-      case PanelEditorTabId.Query:
-        return panel.targets.length;
-      case PanelEditorTabId.Alert:
-        return panel.alert ? 1 : 0;
-      case PanelEditorTabId.Transform:
-        const transformations = panel.getTransformations() ?? [];
-        return transformations.length;
-    }
+  const activeTab = tabs.find((item) => item.active)!;
 
+  if (tabs.length === 0) {
     return null;
-  };
+  }
 
-  render() {
-    const { dashboard, onChangeTab, tabs, panel } = this.props;
-    const styles = getPanelEditorTabsStyles();
-    const activeTab = tabs.find((item) => item.active)!;
+  return (
+    <div className={styles.wrapper}>
+      <TabsBar className={styles.tabBar}>
+        {tabs.map((tab) => {
+          return (
+            <Tab
+              key={tab.id}
+              label={tab.text}
+              active={tab.active}
+              onChangeTab={() => onChangeTab(tab)}
+              icon={tab.icon as IconName}
+              counter={getCounter(panel, tab)}
+            />
+          );
+        })}
+      </TabsBar>
+      <TabContent className={styles.tabContent}>
+        {activeTab.id === PanelEditorTabId.Query && <PanelEditorQueries panel={panel} queries={panel.targets} />}
+        {activeTab.id === PanelEditorTabId.Alert && <AlertTab panel={panel} dashboard={dashboard} />}
+        {activeTab.id === PanelEditorTabId.Transform && <TransformationsEditor panel={panel} />}
```

Comment:
Could also map tab id to a render function and keep that map outside of component:
```
const tabMap = { [PanelEditorTabId.Query]: ({panel}) => <PanelEditorQueries panel={panel} queries={panel.targets} }
```
And then call it `tabMap[activeTab.id]({panel})`. 
Makes the render code a bit cleaner and easier to extend later imo.

Reply:
yes, good suggestion.